### PR TITLE
Fixes #414 and WIP toward Fixing #366 and partially fix Issue #396

### LINKED ISSUE #414: Add seed and iteration parameter to SVD

### PR DESCRIPTION
**Pull request by pseudotensor**

### CODE REVIEW

File: src/interface_py/h2o4gpu/solvers/truncated_svd.py [hunk 1]
```diff
@@ -23,19 +23,33 @@ class TruncatedSVDH2O(object):
                            Either “cusolver” (similar to ARPACK)
                            or “power” for the power method.
 
+    :param: int n_iter: number of iterations (only relevant for power method)
+            Should be large number to run indefinitely
+
+    :param: int random_date: seed (None for auto-generated)
```
Comment (navdeep-G) [inline src/interface_py/h2o4gpu/solvers/truncated_svd.py]: Typo. Should be `random_state`

File: src/interface_py/h2o4gpu/solvers/truncated_svd.py [hunk 2]
```diff
@@ -82,10 +96,15 @@ def fit_transform(self, X, y=None):
...
         param.tol = self.tol
+        param.n_iter = self.n_iter
+        param.random_state = self.random_state
+        param.verbose = 1 if self.verbose else 0
         param.gpu_id = self.gpu_id
 
-        if param.tol <= 0.0:
-            raise ValueError("The `tol` parameter must be > 0.0")
+        if param.tol < 0.0:
+            raise ValueError("The `tol` parameter must be >= 0.0")
```
Comment (navdeep-G) [inline src/interface_py/h2o4gpu/solvers/truncated_svd.py]: Tol should be > 0.0

File: src/interface_py/h2o4gpu/solvers/truncated_svd.py [hunk 3]
```diff
@@ -248,7 +267,7 @@ class TruncatedSVD(object):
 
     def __init__(self,
                  n_components=2,
-                 algorithm="cusolver",
+                 algorithm="power",
```
Comment (navdeep-G) [inline src/interface_py/h2o4gpu/solvers/truncated_svd.py]: I would leave as `cusolver` to keep consistent with previous version.

File: src/interface_py/h2o4gpu/solvers/truncated_svd.py (hunk 2 above)
Comment (pseudotensor) [reply src/interface_py/h2o4gpu/solvers/truncated_svd.py]: It's on purpose.  They can choose 0.0 if they want it to run effectively forever, like explained in params.

File: src/interface_py/h2o4gpu/solvers/truncated_svd.py (hunk 3 above)
Comment (pseudotensor) [reply src/interface_py/h2o4gpu/solvers/truncated_svd.py]: If power is going to be generally better, should use that as default.  Can't expect users to change defaults.

File: src/interface_py/h2o4gpu/solvers/truncated_svd.py (hunk 1 above)
Comment (pseudotensor) [reply src/interface_py/h2o4gpu/solvers/truncated_svd.py]: Thanks.

**Comment by pseudotensor** [comment]
@navdeep-G Here is what has a  nan:

```
h2o4gpu tsvd run
Total time for h2o4gpu tsvd is 0.01196599006652832
h2o4gpu tsvd Singular Values
[250.67248535   0.        ]
h2o4gpu tsvd Components (V^T)
[[0.14196943 0.14243111 0.14300583 0.13927637 0.14067133 0.14231789
  0.14222175 0.14159958 0.14038227 0.13927117 0.13995984 0.13993007
  0.1426447  0.1399776  0.14397652 0.14034724 0.14050162 0.1417561
  0.14095829 0.14117518 0.14354251 0.1411379  0.14120567 0.14171125
  0.14050955 0.14136457 0.14034872 0.14229235 0.1444754  0.14283538
  0.13969038 0.14083743 0.14261284 0.14109325 0.1421081  0.14423983
  0.14076583 0.14055614 0.14123496 0.13993134 0.14167024 0.13971879
  0.1415036  0.14170535 0.13921899 0.14154966 0.14010049 0.14270718
  0.14346381 0.14226256]
 [       nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan]]
h2o4gpu tsvd Explained Variance
[0.08407421 0.06901251]
h2o4gpu tsvd Explained Variance Ratio
[0.02022091 0.01659838]


sklearn run
Total time for sklearn is 0.013918638229370117
Sklearn Singular Values
[250.67248963  22.26525052]
Sklearn Components (V^T)
[[ 0.14196948  0.14243109  0.14300587  0.13927635  0.14067131  0.14231787
   0.14222176  0.14159958  0.14038233  0.1392712   0.13995981  0.13993007
   0.14264471  0.13997766  0.1439765   0.14034722  0.14050159  0.1417561
   0.1409583   0.1411752   0.1435425   0.14113787  0.14120567  0.14171125
   0.14050955  0.1413646   0.14034869  0.14229241  0.14447542  0.14283542
   0.13969041  0.14083743  0.14261282  0.14109324  0.14210811  0.14423986
   0.1407658   0.14055614  0.14123494  0.13993132  0.14167024  0.13971881
   0.14150364  0.14170534  0.13921897  0.14154969  0.14010053  0.14270722
   0.14346384  0.14226256]
 [ 0.11712282  0.17900278 -0.06560087  0.05482444 -0.15732129  0.06845241
   0.02928566 -0.09899839 -0.02174216 -0.04304371 -0.01909575  0.22457317
   0.03071741 -0.08184205  0.04730319  0.28910454 -0.25506042 -0.01042499
   0.03333875  0.10792759  0.08289915 -0.23082418  0.0740953  -0.06619307
   0.08545595 -0.37389742  0.13110945  0.10195325  0.11094715 -0.01633165
  -0.13652009 -0.06531512  0.05577624 -0.06920954 -0.09464793 -0.27038872
  -0.3197246   0.14256088  0.18408372  0.24040472 -0.1829273  -0.08459606
  -0.08200496  0.14839738 -0.00051397 -0.03024335 -0.04045249  0.10710413
   0.06889189  0.09903009]]

```

**Comment by pseudotensor** [comment]
```
tests_open/svd/test_tsvd_bench.py::test_tsvd_error_k5 SVD on 500000 by 100 matrix
Original X Matrix
[[0.19151945 0.62210877 0.43772774 ... 0.81920207 0.05711564 0.66942174]
 [0.76711663 0.70811536 0.79686718 ... 0.10310444 0.80237418 0.94555324]
 [0.97903882 0.88123225 0.62768192 ... 0.82215979 0.62796507 0.11792306]
 ...
 [0.31773671 0.62527382 0.58866184 ... 0.87304627 0.13717136 0.28000008]
 [0.77962647 0.93798681 0.99846925 ... 0.14616526 0.42627708 0.94164386]
 [0.66703042 0.9340073  0.77347522 ... 0.27247573 0.23986564 0.53331485]]


h2o4gpu tsvd run
Total time for h2o4gpu tsvd is 0.6790354251861572
h2o4gpu tsvd Singular Values
[3541.61450195    0.            0.            0.            0.        ]
h2o4gpu tsvd Components (V^T)
[[0.10002381 0.09996234 0.09995778 0.10010591 0.1000362  0.10000781
  0.09993881 0.09989155 0.09994844 0.10009401 0.09997676 0.09991338
  0.09992535 0.0999381  0.10004885 0.10001038 0.10015671 0.10001526
  0.10016877 0.09997884 0.10011817 0.10004561 0.09994294 0.09986766
  0.10000329 0.09995866 0.09994347 0.09997346 0.10007129 0.0999036
  0.10008524 0.1000131  0.10003185 0.10001048 0.09997862 0.10004354
  0.09997182 0.09992619 0.09986454 0.10006943 0.09978941 0.10003491
  0.09994646 0.10001729 0.1000025  0.10001123 0.10000704 0.09995592
  0.09992315 0.10009161 0.10012327 0.09998558 0.10003188 0.09996752
  0.09995818 0.09998848 0.09992085 0.10014058 0.09985561 0.09998788
  0.0999494  0.10010172 0.10017264 0.10002732 0.10003568 0.09999733
  0.10005731 0.10006276 0.10001228 0.10006762 0.1000191  0.09995933
  0.10011411 0.10009572 0.10014059 0.09999558 0.09988362 0.10016145
  0.09991603 0.09985507 0.10010148 0.09986185 0.09997825 0.1000219
  0.10007452 0.09996291 0.10000127 0.10004679 0.09993626 0.09992731
  0.09998287 0.10004407 0.09998161 0.09998313 0.09980367 0.10012543
  0.0999556  0.10008066 0.09997449 0.09986611]
 [       nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan]
 [       nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan        nan        nan
         nan        nan        nan        nan]
:

```

**Comment by pseudotensor** [comment]
only happens when running pytest -n 1 for all tests, as if svd is dependent on some setting that itself is not setting.

**Comment by pseudotensor** [comment]
jenkins and mr-dl10 get it: http://mr-0xc1:8080/blue/organizations/jenkins/h2o4gpu-nccl-cuda8/detail/PR-417/5/pipeline

**Comment by pseudotensor** [comment]
I tried running some reduced set of tests together, and so far no issues:
```
pytest -s -v tests_open/svd tests_open/glm
...
============= 79 passed, 10 skipped, 4 warnings in 409.70 seconds ==============
```


```
pytest -s -v tests_open/svd tests_open/kmeans
...
========================== 25 passed in 36.78 seconds ==========================
...

**Comment by pseudotensor** [comment]
On home computer passed:
```
========================================================================================= 111 passed, 10 skipped, 22 warnings in 455.87 seconds ==========================================================================================
# Test R package
/usr/bin/R-3.1.0 -e 'devtools::test("src/interface_r")'
/bin/bash: /usr/bin/R-3.1.0: No such file or directory
Makefile:468: recipe for target 'dotest' failed
make: *** [dotest] Error 127
jon@pseudotensor:~/h2o4gpu$ 

```

**Comment by pseudotensor** [comment]
Doing this also fails: pytest -n 1 -s -v tests_open/gbm tests_open/glm tests_open/kmeans tests_open/svd tests_open/system

**Comment by pseudotensor** [comment]
This is enough to trigger the problem:

pytest -n 1 -s -v tests_open/glm/test_glm_sklearn.py tests_open/svd

Even this is enough:

pytest -n 1 -s -v tests_open/glm/test_glm_sklearn.py::TestGlmSklearn::test_glm_sklearn_gpu_data0     tests_open/svd/test_tsvd.py::test_tsvd_error_k2
